### ISSUE TITLE
Stuck on big difference files

### ISSUE THREAD

**Issue by bararchy** [MakeNowJust/crystal-diff#5]
For example: 
```crystal
require "diff"
require "colorize"

File.open("/dev/urandom") do |f|
  buff1 = Bytes.new(4096 * 2)
  buff2 = Bytes.new(4096 * 2)

  f.read(buff1)
  f.read(buff2)

  Diff.diff(String.new(buff1), String.new(buff2)).each do |chunk|
  print chunk.data.colorize(
    chunk.append? ? :green : chunk.delete? ? :red : :dark_gray)
end
puts
  
end
```

Will run for a LONG time

**Comment by makenowjust** [comment]
Yes, it is Wu's algorithm feature. The order of this algorithm is `O(NP)` and `P` becomes big in proportion to edit distance size between target texts. So, it slow downs in this case, but it is very fast on small difference files.

It is hard to avoid this issue.

**Comment by bararchy** [comment]
I see, @MakeNowJust would you think a feature to use specific algortihm for the diff can be added? so that maybe we can switch between use-cases? 

`Diff#diff(String, String, algorithm: Algorithm)`

```crystal
enum Algorithm
  Wu
  Myers
end
```

etc..

**Comment by bararchy** [comment]
Adding Myers for example seems quite easy: 
* https://github.com/mattsan/ruby_myers_diff/blob/master/lib/myers_diff.rb  
* https://blog.jcoglan.com/2017/04/25/myers-diff-in-linear-space-implementation/

**Comment by makenowjust** [comment]
I think, simply using Myers' algorithm doesn't solve this. Because [jsdiff](https://github.com/kpdecker/jsdiff) has the same issue and it uses Myers'.

Ruby's `diff-lcs` library and Python's `difflib` work fine for this issue example. `diff-lcs` uses McIlroy-Hunt algorithm and `difflib` uses Gestalt approach. I'm learning them and trying to implement.

**Comment by makenowjust** [comment]
Note: don't need to convert `Bytes` to `String` before passing `Diff.diff`, it works for any collections. In addition, `String` converts invalid unicode sequence and `String#[]` is slow.

**Comment by makenowjust** [comment]
I said "simply using Myers' algorithm"... I mean, we can apply approximation algorithm to Myers' for avoiding worst case and Git's diff implementation does such a thing. But it is complex and hard to implement to me.

**Comment by bararchy** [comment]
@MakeNowJust It seems Myer's Algo is better then LCS -> https://stackoverflow.com/a/42741558/1526733 (based on this answer) in memory consumption and in speed.

**Comment by bararchy** [comment]
@MakeNowJust regarding the Bytes\String, it was just an example, my real use case is only String and it's of curse not from `/dev/urandom` :), it's just large HTTP responses

**Comment by makenowjust** [comment]
I implemented Myers O(ND) algorithm with applying Hirschberg trick and it will solves this issue. I am refactoring it. Coming soon.

However Hirschberg trick can apply some algorithms using edit-path graph, so we can consider Wu O(NP) algorithm with Hirschberg trick and I think it may be faster.

**Comment by bararchy** [comment]
Great news! thanks for the help @MakeNowJust :heart:

**Comment by bararchy** [comment]
@MakeNowJust Any update? really waiting for this awesome feature!

**Comment by makenowjust** [comment]
@bararchy I released [v1.1.0](https://github.com/MakeNowJust/crystal-diff/releases/tag/v1.1.0) now. It doesn't use Myers' algorithm (because it is slower than Wu for less difference files really), and it optimized edit script construction. It is still slow on big difference files however it is faster than now surely.
